File: Days 11 - 20/Day 15/RandomNumberFromStream.cpp
#include <iostream>
#include <random>

int SelectRandomFromStream(int currentNumber, bool resetStream = false)
{
	static int randomStreamNumber = 0;
	static int count = 0;

	if (resetStream)
	{
		randomStreamNumber = 0;
		count = 0;
	}

	count++;

	if (count == 1)
	{
		randomStreamNumber = currentNumber;
	}
	else
	{
		static std::random_device randomSeed;
		static std::mt19937 randomEngine(randomSeed());
		std::uniform_int_distribution<int> randomGenerator(0, count);

		int randomNumber = randomGenerator(randomEngine);

		if (randomNumber == count - 1)
		{
			randomStreamNumber = currentNumber;
		}
	}

	return randomStreamNumber;
}

int main(int argc, char* argv[])
{
	int randomNumber = 0;

	do
	{
		std::cout << "Enter a number (or \"quit\" to stop): ";

		int currentNumber;
		std::cin >> currentNumber;

		if (!std::cin.fail())
		{
			randomNumber = SelectRandomFromStream(currentNumber);
		}
	}
	while (!std::cin.fail());

	std::cin.clear();
	std::cin.ignore(10000, '\n');

	std::cout << "Random number: " << randomNumber << std::endl;

	std::cin.get();

	return 0;
}

File: Days 011 - 020/Day 15/RandomNumberFromStream.cs
using System;

namespace DailyCodingProblem
{
	internal class Day15
	{
		private static Random rng = new Random();

		private static int randomStreamNumber = 0;
		private static int count = 0;

		private static int Main(string[] args)
		{
			int randomNumber = 0;

			Console.Write("Total number of numbers to enter: ");
			int numberCount = Convert.ToInt32(Console.ReadLine());

			for (int i = 0; i < numberCount; i++)
			{
				Console.Write("Enter a number: ");
				int currentNumber = Convert.ToInt32(Console.ReadLine());

				randomNumber = SelectRandomFromStream(currentNumber);
			}

			Console.WriteLine($"Random number: {randomNumber}");

			Console.ReadLine();

			return 0;
		}

		private static int SelectRandomFromStream(int currentNumber, bool resetStream = false)
		{
			if (resetStream)
			{
				randomStreamNumber = 0;
				count = 0;
			}

			count++;

			if (count == 1)
			{
				randomStreamNumber = currentNumber;
			}
			else
			{
				int randomNumber = rng.Next(0, count + 1);

				if (randomNumber == count - 1)
				{
					randomStreamNumber = currentNumber;
				}
			}

			return randomStreamNumber;
		}
	}
}